### PR TITLE
Favorite Pages: Allow User Cross-Site Page Editing

### LINKED ISSUE #24992: Favorite Pages: Allow User Cross-Site Page Editing
### Problem Statement
In pages portlet when you try to cross-site edit a page from a bookmark user can go to edit page but the CSS are broken.

### Steps to Reproduce
1. Create another site
2. Navigate to Pages portlet
3. From site A add a bookmark
4. Switch to site B
5. From nookmarks, try to edit the page
6. See how the CSS are broken

### Acceptance Criteria
User should be able to edit a boomarked page from any site

### External Links
N/A

### Assumptions & Initiation Needs
If the page we try to edit is in a diferent site that the current site, we need to change the current site before go to edit the page.

Yes, we're changing the current site under the hood.

### Quality Assurance Notes & Workarounds
N/A

### PR DESCRIPTION
**Pull request by rjvelazco**
### Proposed Changes
* Switch to the right `site` when going to `edit page`

### Checklist
- [x] Tests
- [x] Translations

### <samp>🤖 Generated by Copilot at 3d1dd4b</samp>

Refactored the `DotEditPageResolver` service and the `DotToolbarComponent` to enable editing pages from different sites. Added and modified methods in the `SiteService` and its mock class to handle site-switching logic. Removed unused import and provider from the `dot-pages.module.ts` file.

## Related Issue
Fixes #24992 

# Explanation of Changes

### <samp>🤖 Generated by Copilot at 3d1dd4b</samp>

*  Refactor `resolve` method of `DotEditPageResolver` class to use `siteService` and extract logic into separate methods ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-4fac9401f4a6c5cf837e65f987b43c3aa95d70cc7144e83484321e5068427546L28-R49), [link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-4fac9401f4a6c5cf837e65f987b43c3aa95d70cc7144e83484321e5068427546R98-R123))
* Add `setSite` and `getPageRenderState` methods to `DotEditPageResolver` class to handle site switching and page state requests ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-4fac9401f4a6c5cf837e65f987b43c3aa95d70cc7144e83484321e5068427546R98-R123))
* Add `Site` and `SiteService` imports to `dot-edit-page-resolver.service.ts` file to access site information and operations ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-4fac9401f4a6c5cf837e65f987b43c3aa95d70cc7144e83484321e5068427546L7-R11))
* Add `switchSiteById` and `switchSite` methods to `SiteService` class to make HTTP requests and update current site ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-ebec7159cb4a8660492c16c6797f79029bb5fedea93d4bbad8f44c84729ca61cL151-R182))
* Add `switchMap` and `tap` imports to `site.service.ts` file to use operators in site switching methods ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-ebec7159cb4a8660492c16c6797f79029bb5fedea93d4bbad8f44c84729ca61cL5-R5))
* Modify `switchToDefaultSite` method of `SiteService` class to use `switchMap` operator instead of subscribing to observable ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-ebec7159cb4a8660492c16c6797f79029bb5fedea93d4bbad8f44c84729ca61cL73-R77))
* Modify `getSiteById` method of `SiteService` class to use correct JSDoc syntax for parameter and return types ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-ebec7159cb4a8660492c16c6797f79029bb5fedea93d4bbad8f44c84729ca61cL135-R137))
* Modify `siteChange` method of `DotToolbarComponent` class in `dot-toolbar.component.ts` file to subscribe to `siteService.switchSite` observable ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-702d70cf5d8b2d5e27374f9e33e704440cff80ff92bef64757295c2bcc6270faL42-R42))
* Remove unused and redundant `SiteService` import and provider from `dot-pages.module.ts` file ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-47e3295a5749aab684a0a476c5795118c8b83676b76c9e1aa3e5d4b365d2c6c5L22), [link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-47e3295a5749aab684a0a476c5795118c8b83676b76c9e1aa3e5d4b365d2c6c5L54))
* Add `of` import to `site.service.mock.ts` file to create observable from value in `switchSite` method ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-2252907b5ad0200016c34033f4cf8160e4c32df51e02c4b85460912179a597c2L1-R3))
* Add `switchSiteById` and `switchSite` methods to `SiteServiceMock` class in `site.service.mock.ts` file to mimic real `siteService` methods for testing ([link](https://github.com/dotCMS/core/pull/25005/files?diff=unified&w=0#diff-2252907b5ad0200016c34033f4cf8160e4c32df51e02c4b85460912179a597c2L40-R48))

### Screenshots

https://github.com/dotCMS/core/assets/72418962/36ad83c0-f291-4083-be40-0c20c64f9f67

### CODE REVIEW

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts [hunk 1]
```diff
@@ -147,22 +148,38 @@ export class SiteService {
...
+        return this.getSiteById(id).pipe(
+            switchMap((site) => this.switchSite(site)),
+            take(1)
+        );
+    }
+
     /**
      * Change the current site
      * @param Site site
      * @memberof SiteService
```
Comment (alfredo-dotcms) [inline core-web/libs/dotcms-js/src/lib/core/site.service.ts:168]: fix doc `return`

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/resolvers/dot-experiment-experiment.resolver.spec.ts [hunk 2]
```diff
@@ -66,6 +73,7 @@ describe('DotExperimentExperimentResolver', () => {
                 DotFavoritePageService,
                 { provide: DotMessageDisplayService, useClass: DotMessageDisplayServiceMock },
                 { provide: DotRouterService, useClass: MockDotRouterService },
+                { provide: SiteService, useValue: siteServiceMock },
```
Comment (zJaaal) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/resolvers/dot-experiment-experiment.resolver.spec.ts:76]: Same question

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts [hunk 3]
```diff
@@ -104,4 +88,43 @@ export class DotEditPageResolver implements Resolve<DotPageRenderState> {
...
+
+    private setSite(id: string): Observable<Site> {
+        const currentSiteId = this.siteService.currentSite?.identifier;
+        const shouldSwitchSite = id && id !== currentSiteId;
+
+        // If we have a site id and is different from the current one, we switch
+        return shouldSwitchSite
+            ? this.siteService.switchSiteById(id).pipe(
+                  catchError((err: HttpErrorResponse) => {
+                      console.warn(err);
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts:100]: use `this.dotHttpErrorManagerService.handle(err, true)` so in that way the user can know something weird happened, in an unobtrusive way (toast message)

**Review by zJaaal — COMMENTED** [review]
.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts [hunk 4]
```diff
@@ -25,37 +25,21 @@ export class DotEditPageResolver implements Resolve<DotPageRenderState> {
...
+        private dotHttpErrorManagerService: DotHttpErrorManagerService,
+        private siteService: SiteService
     ) {}
 
     resolve(route: ActivatedRouteSnapshot): Observable<DotPageRenderState> {
         const data = this.dotPageStateService.getInternalNavigationState();
+        const renderOptions = this.getDotPageRenderOptions(route);
+        const currentSection = route.children[0].url[0].path;
+        const isLayout = currentSection === 'layout';
+        const hostId = route.queryParams?.host_id;
```
Comment (alfredo-dotcms) [inline core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts:37]: what happens if this is `null` ?
would it break on `line 98` ???

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts [hunk 5]
```diff
@@ -70,10 +70,11 @@ export class SiteService {
         if (siteIdentifier === this.selectedSite.identifier) {
             name === 'ARCHIVE_SITE'
                 ? this.switchToDefaultSite()
-                      .pipe(take(1))
-                      .subscribe((currentSite: Site) => {
-                          this.switchSite(currentSite);
-                      })
+                      .pipe(
+                          take(1),
+                          switchMap((site) => this.switchSite(site))
```
Comment (alfredo-dotcms) [inline core-web/libs/dotcms-js/src/lib/core/site.service.ts:75]: why this change?

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts [hunk 6]
```diff
@@ -147,22 +148,38 @@ export class SiteService {
             );
     }
 
+    /**
+     * Change the current Site by site identifier
+     * @param Site site
```
Comment (alfredo-dotcms) [inline core-web/libs/dotcms-js/src/lib/core/site.service.ts:153]: fix doc

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts (hunk 4 above)
Comment (rjvelazco) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.ts:37]: If the `hostId` is equal to `null` or `undefined`, we don't switch the `site`. There is a test for that case: 

https://github.com/dotCMS/core/blob/b6d62b3757669ad289d11d67dcedefef5aacbc8c/core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.spec.ts#L166-L176

File: core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.spec.ts [hunk 7]
```diff
@@ -72,6 +76,7 @@ describe('DotEditPageResolver', () => {
                 DotFavoritePageService,
                 { provide: DotRouterService, useClass: MockDotRouterService },
                 { provide: DotMessageDisplayService, useClass: DotMessageDisplayServiceMock },
+                { provide: SiteService, useValue: siteServiceMock },
```
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-edit-page/shared/services/dot-edit-page-resolver/dot-edit-page-resolver.service.spec.ts:79]: Not a specific reason. I'll use `useClass` that way is better.

File: core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/resolvers/dot-experiment-experiment.resolver.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjvelazco) [reply core-web/apps/dotcms-ui/src/app/portlets/dot-experiments/shared/resolvers/dot-experiment-experiment.resolver.spec.ts:76]: Done!

**Review by manuelrojas — APPROVED** [review]

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts (hunk 5 above)
Comment (rjvelazco) [reply core-web/libs/dotcms-js/src/lib/core/site.service.ts:75]: That function now returns an observable. That's why I use it inside a `switchMap`

**Review by zJaaal — APPROVED** [review]

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts (hunk 1 above)
Comment (rjvelazco) [reply core-web/libs/dotcms-js/src/lib/core/site.service.ts:176]: Done!

**Review by alfredo-dotcms — APPROVED** [review]

**Pull request merged**